Userland/chmod.cpp

```diff
@@ -3,151 +3,143 @@
+bool string_to_mode(char access_scope, const char*& access_string, mode_t& mode);
+bool apply_permission(char access_scope, char permission, char operation, mode_t& mode);
 
 int main(int argc, char** argv)
 {
-    if (argc != 3) {
-        printf("usage: chmod <octal-mode> <path>\n"
-               "       chmod [ugoa...][+-=][rwx...][,...] <path>\n");
+    if (argc < 3) {
+        printf("usage: chmod <octal-mode> <path,...>\n"
+               "       chmod [[ugoa][+-=][rwx...],...] <path,...>\n");
         return 1;
     }
 
-    unsigned mode = 0;
-    int rc = 0;
+    mode_t mode;
+    bool ret_success = true;
 
     if(argv[1][0] >= '0' && argv[1][0] <= '7') {
-        rc = sscanf(argv[1], "%o", &mode);
-        if (rc != 1) {
+        if (sscanf(argv[1], "%ho", &mode) != 1) {
             perror("sscanf");
             return 1;
     	}
     } else {
-        char* access_str = argv[1];
-        struct stat curr_access;
+        const char* access_string = argv[1];
+        struct stat current_access;
 
-        rc = stat(argv[2], &curr_access);
-        if(rc != 0) {
+        if(stat(argv[2], &current_access) != 0) {
             perror("stat");
             return 1;
         }
-        mode = curr_access.st_mode;
-        while(*access_str != '\0') {
-            switch(*access_str) {
-                case 'u': rc = str_to_octal_mode('u', &(++access_str), mode); break;
-                case 'g': rc = str_to_octal_mode('g', &(++access_str), mode); break;
-                case 'o': rc = str_to_octal_mode('o', &(++access_str), mode); break;
-                case 'a': rc = str_to_octal_mode('a', &(++access_str), mode); break;
+        mode = current_access.st_mode;
+        while(*access_string != '\0') {
+            switch(*access_string) {
+                case 'u': ret_success = string_to_mode('u', ++access_string, mode); break;
+                case 'g': ret_success = string_to_mode('g', ++access_string, mode); break;
+                case 'o': ret_success = string_to_mode('o', ++access_string, mode); break;
+                case 'a': ret_success = string_to_mode('a', ++access_string, mode); break;
                 case '=':
                 case '+':
-                case '-': rc = str_to_octal_mode('a', &access_str, mode);     break;
-                case ',': ++access_str;                                       break;
-                default: rc = -1;
+                case '-': ret_success = string_to_mode('a', access_string, mode);   break;
+                case ',': ++access_string;                                          break;
+                default: ret_success = false;
             }
-            if(rc != 0) {
+            if(!ret_success) {
                 fprintf(stderr, "chmod: invalid mode: %s\n", argv[1]);
        	        return 1;
             }
         }
     }
 
-    rc = chmod(argv[2], mode);
-    if (rc < 0) {
-        perror("chmod");
-        return 1;
+    int i = 2;
+    while(i < argc) {
```

Comment:
This doesn't look right. We compute `mode` for the first file, and then set every file to that exact mode.
What happens if I say `chmod u+x *` for instance? :)

Reply:
In this case permissions will change to `u+x` for each file in the current directory :) I thought, it is right behaviour of `chmod` with multiple files...according to `man chmod` and behaviour of `chmod` in my Linux distro. I think, it is better to say that `chmod` sets the same `mode` for all files in the set.

Reply:
But `mode` is computed based on `argv[2]` above. So when using the + or - modifier, the original mode from `argv[2]` will be transferred to all subsequent arguments as well.

Here's how my host system behaves:
```
$ touch file1 file2 file3
$ chmod 600 file1
$ chmod 660 file2
$ chmod 666 file3
$ chmod u-r *
$ ls -l 
--w------- 1 kling 0 Jul 12 14:01 file1
--w-rw---- 1 kling 0 Jul 12 14:01 file2
--w-rw-rw- 1 kling 0 Jul 12 14:01 file3
```

Your approach would make all three files be `--w-------`, which is not what we want.

Reply:
Oh, i understood :) Yes, it's mistake, i get current `mode` using `stat` call for argv[2]. I will fix it!